sdk/core/azure-core/azure/core/serialization.py

```diff
@@ -29,6 +37,166 @@ def __bool__(self) -> bool:
+
+        - If the condition is a type, we add it to a type map.
+        - If the condition is a callable function, we add it to a predicate list.
+
+        The handler function is expected to take two arguments: the target type and the data dictionary,
+        and return an instance of the target type.
+
+        Examples:
+
+        .. code-block:: python
+
+            @registry.register_deserializer(CustomModel)
+            def deserialize_single_type(cls: Type[CustomModel], data: dict) -> CustomModel:
+                return cls(**data)
+
+            @registry.register_deserializer(lambda t: issubclass(t, BaseModel))
+            def deserialize_with_condition(cls: Type[BaseModel], data: dict) -> BaseModel:
+                return cls(**data)
+
+            # Called manually for a specific type
+            def custom_deserializer(cls: Type[CustomModel], data: Dict[str, Any]) -> CustomModel:
+                return cls(custom=data["custom"])
+
+            registry.register_deserializer(CustomModel)(custom_deserializer)
+
+        :param condition: A type or a callable predicate function that takes an object and returns a bool.
+        :type condition: Union[Type, Callable[[Any], bool]]
+        :return: A decorator that registers the handler function.
+        :rtype: Callable
+        :raises TypeError: If the condition is neither a type nor a callable.
+        """
+
+        def decorator(handler_func: Callable[[Type, Dict[str, Any]], Any]) -> Callable[[Type, Dict[str, Any]], Any]:
+            if isinstance(condition, type):
+                self._deserializer_types[condition] = handler_func
+            elif callable(condition):
+                self._deserializer_predicates.append((condition, handler_func))
+            else:
+                raise TypeError("Condition must be a type or a callable predicate function.")
+
+            self._deserializer_cache.clear()
+            return handler_func
+
+        return decorator
+
+    def get_serializer(self, obj: Any) -> Optional[Callable]:
+        """Gets the appropriate serializer for an object.
+
+        It first checks the type dictionary for a direct type match.
+        If no match is found, it iterates through the predicate list to find a match.
+
+        Results of the lookup are cached for performance based on the object's type.
+
+        :param obj: The object to serialize.
+        :type obj: any
+        :return: The serializer function if found, otherwise None.
+        :rtype: Optional[Callable]
+        """
+        obj_type = type(obj)
+        if obj_type in self._serializer_cache:
+            return self._serializer_cache[obj_type]
+
+        handler = self._serializer_types.get(type(obj))
+        if not handler:
+            for predicate, pred_handler in self._serializer_predicates:
+                if predicate(obj):
+                    handler = pred_handler
+                    break
+
+        self._serializer_cache[obj_type] = handler
```

Comment:
I think there might be a chance where handler is never set, you can't find it in `self._serializer_predicates`. I think it would be safer to throw a warning if handler is still null here before setting serializer_cache